src/osmet/mod.rs

```diff
@@ -0,0 +1,579 @@
+// Copyright 2020 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::fs::{File, OpenOptions};
+use std::io::{self, Seek, SeekFrom, Write};
+use std::os::unix::fs::FileTypeExt;
+use std::path::{Path, PathBuf};
+
+use error_chain::bail;
+use nix::mount;
+use serde::{Deserialize, Serialize};
+use walkdir::WalkDir;
+use xz2::write::XzEncoder;
+
+use crate::blockdev::*;
+use crate::cmdline::*;
+use crate::errors::*;
+
+mod fiemap;
+mod file;
+mod io_helpers;
+mod unpacker;
+
+use crate::osmet::fiemap::*;
+use crate::osmet::io_helpers::*;
+use crate::osmet::file::*;
+use crate::osmet::unpacker::*;
+
+// just re-export OsmetUnpacker
+pub use crate::osmet::unpacker::OsmetUnpacker;
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Mapping {
+    extent: Extent,
+    object: Sha256Digest,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct OsmetPartition {
+    start_offset: u64,
+    end_offset: u64,
+    mappings: Vec<Mapping>,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct Osmet {
+    /// Partitions for which we've registered mappings.
+    partitions: Vec<OsmetPartition>,
```

Comment:
At the moment, `find_matching_osmet_in_dir` needs to load all the mappings before it can check the header for the desired values.  Would it make sense to restructure the osmet file as a header plus a separately-serialized body?

Reply:
Yeah, good call. Goes well with the other header rework mentioned earlier.